Sources tests are failing

Looks like the sources tests are now failing: https://github.com/DNSCrypt/dnscrypt-proxy/runs/1742240545

I think some of that was due to relay randomization, but the rest may have been caused by the recent changes to keep updated lists in memory.

Do you have more insights on what's happening @ianbashford @welwood08 ?

I think I got how the error happens.

First:
![shuffle_issue_in_tests](https://user-images.githubusercontent.com/19585474/105488461-99f5eb80-5cec-11eb-9492-c28ad2c62918.png)
The directory part of path indicates the `state`.
https://github.com/DNSCrypt/dnscrypt-proxy/blob/0ab9e30fa9a7479e46f90bfdd0b6ce49119d1560/dnscrypt-proxy/sources_test.go#L33-L45

Then, let's take an example:
The error:
```
    --- FAIL: TestNewSource/cache_missing,_download_missing-sig,correct/10-1 (0.01s)
        sources_test.go:231: Unexpected HTTP request log
            Checker:  DeepEqual
            Expected: (map[string]uint) (len=4) {
              (string) (len=11) "/0/empty.md": (uint) 1,
              (string) (len=19) "/0/empty.md.minisig": (uint) 1,
              (string) (len=11) "/5/empty.md": (uint) 1,
              (string) (len=19) "/5/empty.md.minisig": (uint) 1
            }
            Actual:   (map[string]uint) (len=2) {
              (string) (len=11) "/0/empty.md": (uint) 1,
              (string) (len=19) "/0/empty.md.minisig": (uint) 1
            }
```
And the inputs:
```
id: 10-1, e.urls: [http://127.0.0.1:63635/5/empty.md http://127.0.0.1:63635/0/empty.md]
Before Shuffle source.urls: [http://127.0.0.1:63635/5/empty.md http://127.0.0.1:63635/0/empty.md]
After Shuffle source.urls: [http://127.0.0.1:63635/0/empty.md http://127.0.0.1:63635/5/empty.md]
```

Finally, you get it. The error occurs when a `0` leading path is shuffled to the 1st, which means `TestStateCorrect` and **leads to an end**!

--- a/dnscrypt-proxy/config.go
+++ b/dnscrypt-proxy/config.go
@@ -788,6 +788,9 @@ func (config *Config) printRegisteredServers(proxy *Proxy, jsonOutput bool) erro
 func (config *Config) loadSources(proxy *Proxy) error {
 	for cfgSourceName, cfgSource_ := range config.SourcesConfig {
 		cfgSource := cfgSource_
+		rand.Shuffle(len(cfgSource.URLs), func(i, j int) {
+			cfgSource.URLs[i], cfgSource.URLs[j] = cfgSource.URLs[j], cfgSource.URLs[i]
+		})
 		if err := config.loadSource(proxy, cfgSourceName, &cfgSource); err != nil {
 			return err
 		}

--- a/dnscrypt-proxy/sources.go
+++ b/dnscrypt-proxy/sources.go
@@ -197,9 +197,6 @@ func NewSource(name string, xTransport *XTransport, urls []string, minisignKeySt
 		return source, err
 	}
 	source.parseURLs(urls)
-	rand.Shuffle(len(source.urls), func(i, j int) {
-		source.urls[i], source.urls[j] = source.urls[j], source.urls[i]
-	})
 	if _, err = source.fetchWithCache(xTransport, timeNow()); err == nil {
 		dlog.Noticef("Source [%s] loaded", name)
 	}

